streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -371,19 +372,27 @@ InternalTopologyBuilder builder() {
      * @throws StreamsException if the store's change log does not contain the partition
      */
     boolean updateNewAndRestoringTasks() {
-        active.initializeNewTasks();
-        standby.initializeNewTasks();
+        if (!active.allTasksRunning()) {
```

Comment:
This condition passes if we have any tasks in `created`, `restoring` or `suspended`, but inside the condition we have logic for both moving tasks out of `created` as well as out of `restoring`, which makes it a bit clumsy.

I'd suggest we make the interface finer grained (this is internal interface anyways so we can change it to however we want): breaking the `allTasksRunning` into separate conditions:

1. No tasks in `created` or `restoring`, we can now return true from `updateNewAndRestoringTasks` to transit to RUNNING.

2. Some tasks in `created`: call `initializeNewTasks`.

3. Some tasks in `restoring`: call `changelogReader.restore`.

Reply:
I actually just changed `allTasksRunning` to be true if there are tasks in suspended but also running, but that's still kind of clumsy. I think that suggestion makes sense though, I'll try it out (this method definitely needs more work to really support cooperative, I actually suspect it may be responsible in part for the failing test. we'll see..)